capsules/src/net/tlv.rs

```diff
@@ -0,0 +1,1324 @@
+        buf[0] = (PrefixSubTlvType::from(self) as u8) << 1 | stable_bit;
+        buf[1] = value_width as u8;
+        stream_done!(2)
+    }
+
+    // Returns PrefixSubTlv and true if stable, false otherwise.
+    pub fn decode(buf: &[u8]) -> SResult<Option<(PrefixSubTlv, bool)>> {
+        let (offset, tlv_type_field) = dec_try!(buf; decode_u8);
+        let tlv_type_raw = tlv_type_field >> 1;
+        let tlv_type = PrefixSubTlvType::from(tlv_type_raw);
+        let stable = (tlv_type_field & 1u8) > 0;
+        let (offset, length) = dec_try!(buf, offset; decode_u8);
+        match tlv_type {
+            PrefixSubTlvType::HasRoute => {
+                stream_done!(offset + length as usize,
+                             Some((PrefixSubTlv::HasRoute(&buf[offset..offset + length as usize]),
+                                   stable)))
+            }
+            PrefixSubTlvType::BorderRouter => {
+                stream_done!(offset + length as usize,
+                             Some((PrefixSubTlv::BorderRouter(&buf[offset..
+                                                               offset + length as usize]),
+                                   stable)))
+            }
+            PrefixSubTlvType::SixLoWpanId => {
+                let (offset, first_byte) = dec_try!(buf, offset; decode_u8);
+                let context_id_compress = (first_byte & 0b10000) > 0;
+                let context_id = first_byte & 0b1111;
+                let (offset, context_length) = dec_try!(buf, offset; decode_u8);
+                stream_done!(offset,
+                             Some((PrefixSubTlv::SixLoWpanId {
+                                       context_id_compress: context_id_compress,
+                                       context_id: context_id,
+                                       context_length: context_length,
+                                   },
+                                   stable)))
+            }
+            PrefixSubTlvType::NotPresent => stream_done!(offset, None),
+        }
+    }
+}
+
+#[repr(u8)]
+pub enum PrefixSubTlvType {
+    HasRoute = 0,
+    BorderRouter = 2,
+    SixLoWpanId = 3,
+    NotPresent,
+}
+
+impl From<u8> for PrefixSubTlvType {
+    fn from(tlv_type: u8) -> Self {
+        match tlv_type {
+            0 => PrefixSubTlvType::HasRoute,
+            2 => PrefixSubTlvType::BorderRouter,
+            3 => PrefixSubTlvType::SixLoWpanId,
+            _ => PrefixSubTlvType::NotPresent,
+        }
+    }
+}
+
+impl<'a, 'b> From<&'a PrefixSubTlv<'b>> for PrefixSubTlvType {
+    fn from(prefix_sub_tlv: &'a PrefixSubTlv<'b>) -> Self {
+        match *prefix_sub_tlv {
+            PrefixSubTlv::HasRoute(_) => PrefixSubTlvType::HasRoute,
+            PrefixSubTlv::BorderRouter(_) => PrefixSubTlvType::BorderRouter,
+            PrefixSubTlv::SixLoWpanId { .. } => PrefixSubTlvType::SixLoWpanId,
+        }
+    }
+}
+
+pub struct HasRouteTlvValue {
+    r_border_router_16: u16,
+    r_preference: u8,
```

Comment:
what's the `r_` here, `route`? We tend to prefer verbose names where they make sense